### ISSUE TITLE
spooky type checking of optional fields in an array

### ISSUE THREAD

**Issue by itajaja** [Microsoft/TypeScript#4119]
consider the following piece of code:

````ts
interface A {
	aFoo: number,
}

interface B {
	a?: A
}

// This object fails, because test1[0].a.aFoo is not defined
let test1: B[] = [{
	a: {}
}];

// adding a second object where foo is not defined makes the typechecking pass - WTF?
let test2: B[] = [{
	a: {}
}, {
}];

// adding a[1].a = null makes the typechecking fail again (as axpected, because of a[0].a.aFoo not being defined)
let test3: B[] = [{
	a: {}
}, {
	a: null
}];
````

I would expect the three test objects to fail the type checking, but `test2` passes. It seems that adding a second item to the array where the optional `a:A` is not defined, makes the typechecking for the first object to pass. I tried to reason about why is doing it, but I came to the conclusion this is likely a bug.

playground [here](http://www.typescriptlang.org/Playground#src=interface%20A%20%7B%0A%09aFoo%3A%20number%2C%0A%7D%0A%0Ainterface%20B%20%7B%0A%09a%3F%3A%20A%0A%7D%0A%0A%2F%2F%20This%20object%20fails%2C%20because%20test1%5B0%5D.a.aFoo%20is%20not%20defined%0Alet%20test1%3A%20B%5B%5D%20%3D%20%5B%7B%0A%09a%3A%20%7B%7D%0A%7D%5D%3B%0A%0A%2F%2F%20adding%20a%20second%20object%20where%20foo%20is%20not%20defined%20makes%20the%20typechecking%20pass%20-%20WTF%3F%0Alet%20test2%3A%20B%5B%5D%20%3D%20%5B%7B%0A%09a%3A%20%7B%7D%0A%7D%2C%20%7B%0A%7D%5D%3B%0A%0A%2F%2F%20adding%20a%5B1%5D.a%20%3D%20null%20makes%20the%20typechecking%20fail%20again%20(as%20axpected%2C%20because%20of%20a%5B0%5D.a.aFoo%20not%20being%20defined)%0Alet%20test3%3A%20B%5B%5D%20%3D%20%5B%7B%0A%09a%3A%20%7B%7D%0A%7D%2C%20%7B%0A%09a%3A%20null%0A%7D%5D%3B)

I tried to find a similar issue already opened, but I couldn't find any, sorry if this problem was already discovered.

**Comment by JsonFreeman** [comment]
@itajaja What version of TS are you using? Is it the latest from master? Specifically, I think this would be fixed by #3823, which is after TS 1.5. We used to reduce union types in a way that lost information, and in your case the `{ a: { } }` in test2 would have been erased, but that lossy reduction was removed in the change I mentioned.

**Comment by ahejlsberg** [comment]
This is indeed fixed by #3823, the latest compiler gives an error in all three cases above.

**Comment by itajaja** [comment]
yes I am using 1.5. Thanks for the quick update!
